powermon/commands/response_definition.py

```diff
@@ -8,6 +8,7 @@
 class ResponseType(LowercaseStrEnum):
     ACK = auto()
     INT = auto()
+    TEN_INT = auto()
```

Comment:
I dont want to head this way - I did start off like this for mppsolar logic, but soon ended up with many of this types
eg a quick look finds the below 
```
BigHex2Short:(r-30000)/10
BigHex2Short:r/10
BigHex2Short:r/1000
Hex2Int:r-40
BigHex2Short:r/100
BigHex2Short:(r-2731)/10
BigHex2Short:(r&0x7FFF)/100*(((r&0x8000)>>15)*2-1)
LittleHex2UInt:r/1000
LittleHex2Short:r/1000
```

and the list doesnt include the TEN_INT
Im thinking more of a template type, probably jinja2 where any base type can also have a template applied

Reply:
But this will end up with a lot of extra ResponseTypes
I agree the different value types make sense, but when the answer needs math / transformation having a new ResponseType seems wrong.
I still prefer to have the ability to transform any value. Unit testing is still available with this approach
eg
Response is INT, unit test is straight forward
Response is INT with template/transformation - unit test is still straight forward (just need to supply raw response and template)
Else we end up with ResponseTypes like `BigHex2Short_Minus2731_Divide10` for perhaps a single result

Reply:
How many types will we need to cover the different transformations? If it's less than 20 it's no big deal. And if the number starts to get unmanageable we can change course.

Another advantage of having different types is that if we need to change a type we only need to update one method instead of tracking down all the spots the logic appears.

Reply:
We can also name the response types more descriptively and less literally. What does BigHex2Short_Minus2731_Divide10 do? convert Kelvin to Celsius? Then call it INT_KELVIN_TO_CELSIUS

Reply:
> How many types will we need to cover the different transformations? If it's less than 20 it's no big deal. And if the number starts to get unmanageable we can change course.

Probably about 20, but I dont know tbh
> 
> Another advantage of having different types is that if we need to change a type we only need to update one method instead of tracking down all the spots the logic appears.

But your approach is the opposite, for this example if INT processing needs updating youll need to change both INT and TEN_INT (and 100_INT and 1000_INT ....)

The type is still INT, just to get the display value for some situations we need to divide by 10
Or more realistically the BigHex2Short has logic to convert to a value, but many of the values needs manipulation to give the 'correct' display - example above is to subtract 2731 before dividing by 10, why would you create another type for this single situation?

Reply:
> We can also name the response types more descriptively and less literally. What does BigHex2Short_Minus2731_Divide10 do? convert Kelvin to Celsius? Then call it INT_KELVIN_TO_CELSIUS

First BigHex2Short decode sa 2 byte hexString to int (big endian coded) with `unpack(">h", hexString)[0]` 
And yes, the resulting value is 0.1K in units
There are also 'regular' BigHex2Short responses, and BigHex2Short responses that are 10mV units and some that are mV (or mA) units....

Reply:
I still see a ResponseType and a transformation
INT and no transform
INT and divide by ten
BigHex2Short and no transform
BigHex2Short and divide by 10
BigHex2Short and divide by 10 and convert K to C
BigHex2Short and divide by 100, or 1000 etc

I just dont think the 2 should be combined into 1 type

Reply:
We could do a more complex type structure instead of keeping it flat.

- Response
  - INT Response
    - INT divide 10 Response
  - BigHex2Short
    - BigHex2Short divide by 10
    - BigHex2Short divide by 10 and convert K to C

Reply:
I will work on a pull request to demonstrate tonight (it's 3:50pm my time)

Reply:
why not a single 'divide by 10' transform (which then can be used on INT, FLOAT, BigHex2Short, Hex2Int, LittleHex2Short, BigHex2Float, LittleHex2Float, LittleHex2UInt, LittleHex2Int....)

Reply:
How about we introduce the concept of a unit of measurement. We can define which unit the device is giving us and what unit we want to output. Then we can store how to convert between units at the code level.

So BigHex2Short_Minus2731_Divide10 could be represented as:
 unit_from_device: DECIKELVIN, unit_in_response: CELSIUS

Reply:
we'd need configuration items to specify desired units, and would need to have 10mV and 10mA etc as well as converters 10mV to V, 10mA to A etc. still seems like creating a lot of objects
What about response is 20231019 and we want to get year is 2023 month is 10 etc...
response is Wh and want KWh but only for one of the responses but not the others...

Reply:
How about we pass in a method as a parameter? I think that's possible in python.

Something like UnitHelper.decikelvin_to_celsius Then we can test that logic easily. The type into or bytes will be passed in separately to inform how to decode.

Reply:
getting more complex
ResponseType should be 'how is the data encoded, how do i get a value from it'
The 'this value means deci kelvin or mV, 10mV, 0.1V or mA' is separate - ideally we can convert it to 'standard' units after initially implementing this with a new ResponseType like the original PR I changed it to a template because of the increasing amount of types that were needed. Adding unit specific logic (A and V are different) would increase this even further
Either we go with PR like approach, or a transformation approach
PR-like would be
INT
deka_INT (need to /10 to get base)
deci_INT (need to *10 to get base)
note would also need deka_FLOAT, deka_BigHex2Short etc

temperature maybe should be dealt with separately as some would want C and some F (and maybe some K), so maybe that needs a config item and more complex types

transform would be to pass a transformation 'something' (string etc) that the value is transformed with (maybe a `transform=deka` in extra info)

remember we get responses like
`b"\xa5\x01\x90\x08\x02\x10\x00\x00uo\x03\xbc\xf3"`, with the decode looking like
```
           ["BigHex2Short:r/10", 2, "Battery Bank Voltage", "V"],
            ["BigHex2Short:r/10", 2, "acquistion", "V"],
            ["BigHex2Short:(r-30000)/10", 2, "Current", "A"],
            ["BigHex2Short:r/10", 2, "SOC", "%"],
```
and also like (pi18)
`b"^D0882300,217,2300,500,217,5000,5000,480,500,540,450,552,545,2,10,060,1,1,1,9,1,0,0,0,1,00\r",` 
```
           ["int", "AC output rating active power", "W", {"icon": "mdi:power-plug", "device-class": "power"}],
            ["10int", "Battery rated voltage", "V", {"icon": "mdi:battery-outline", "device-class": "voltage"}],
            ["10int", "Battery re-charge voltage", "V", {"icon": "mdi:battery-outline", "device-class": "voltage"}],
            ["10int", "Battery re-discharge voltage", "V", {"icon": "mdi:battery-outline", "device-class": "voltage"}],
            ["10int", "Battery under voltage", "V", {"icon": "mdi:battery-outline", "device-class": "voltage"}],
```
note this still has the 10int type in use (wonder if it still works...)

Reply:
After sleeping on it I have a couple of thoughts:
1. We should replace the lists with dictionaries so we don't have to rely on the position in the list, should make future changes a bit easier and more readable for people new to the project

        "response": [
            [0, "Time until the end of absorb charging", ResponseType.INT, "sec"],
            [1, "Time until the end of float charging", ResponseType.INT, "sec"]
        ]

becomes

        "response": [
            {"index":0, "description":"Time until the end of absorb charging", "type":ResponseType.INT, "unit":"sec"},
            {"index":1, "description":"Time until the end of float charging", "type":ResponseType.INT, "unit":"sec"}
        ]

2. You are right that the Response data type and the transformation should be separate. We have a few different options:
**2.1.** We could split ResponseType into ResponseDecode and ResponseTransform. Our hypothetical BigHex2Short_Minus2731_Divide10 becomes ResponseDecode.BigHex2Short, ResponseTransform.DECIKELVIN_TO_CELSIUS
For millivolt to volt ResponseTranform could be MILLI, but I think MILLIVOLT_TO_VOLT is way more readable and worth the extra types. I prefer verbose easy to read code over terse code.
**2.2.** If you think there are too many transform types then we could have ResponseDecode, ResponseDeviceUnit, ResponseUnit. Instead of ResponseTransform.MILLIVOLT_TO_VOLT we use ResponseDeviceUnit.MILLIVOLT and ResponseUnit.VOLT. Then the code works out the translation and if it's valid. It is easy to unit test and is still readable without a ton of different types. We would prevent powermon from starting up if an invalid unit combo is presented.

Reply:
I think the dict change is good - also allows the 'extra info' to be removed and added as additional keys
`{"index": 0, ..... "icon": "mdi:question", ...`
I still see ResponseDecode and ResponseType (what is the type of the data in the response) tbh
Probably can get rid of "type" as keyword too (seeing it is a python reserved word)
I guess we have the device response and the response we want for the results as separate concepts
if device response is ResponseXxxx then we should use something else for the other concept
going back to earlier thoughts, if we have ResponseType that allows us to understand the data, we can add a 'what units is the data returned in' and a 'what is the desired (or default) units for the result

ResponseType
ResponseUnits
DesiredUnits ??? (or something much better)

We then do transforms ResponseUnits to DesiredUnits but can also have overides (globally is easy, individually is harder but might be useful)

ie data is \x08, ResponseType is BigHex2Short (or whatever better name works), ResponseUnits are deka_volts, DesiredUnits are Volts, so process with deka_Volts to Volts 
Though this last one is a big if/switch statement

Reply:
[sorry for all the 'stream of thoughts']

I also wonder if we can simplify ACK/SETTERS
response and test_responses are always the same, so can we just remove them and build into the logic?
```
"F": {
        "name": "F",
        "description": "Set Device Output Frequency",
        "help": " -- examples: F50 (set output frequency to 50Hz) or F60 (set output frequency to 60Hz)",
        "response_type": ResultType.ACK,
        "response": [[0, "Command execution", ResponseType.ACK, {"NAK": "Failed", "ACK": "Successful"}]],
        "test_responses": [b"(NAK\x73\x73\r", b"(ACK\x39\x20\r",],
        "regex": "F([56]0)$",
    },
```

Reply:
So you are leaning toward 2.2? 
For serial number then we make ResponseUnit.SERIAL_NUMBER and DesiredUnit.STRING (or whatever) then put that logic in the transform between serial_number and string. If it differs between protocols then it could be ResponseUnit.PI18_SERIAL_NUMBER

Reply:
yes leaning to 2.2
question is best way to determine what transform to apply (including the ability to override it)



> If it differs between protocols then it could be ResponseUnit.PI18_SERIAL_NUMBER

yes this works - custom or specific ResponseUnits can deal with most/all weird situations (as long as we have a good solution to the Q above)

Reply:
I would write objects for each DesiredUnit (OutputUnit? TranslatedUnit? DecodedUnit?) that knows how to convert from the ResponseUnit. There will only be a few types of DesiredUnits and they can hold all that logic.

Instead of needing a way to override, adding a new ResponseUnit and adding corresponding code to the DesiredUnit should work nicely. I really don't think any code should go into the protocols, keeping the code in proper classes will make maintenance easier.

Reply:
> Away from computer so excuse typos
>
>Override is more so an individual could choose to use Fahrenheit instead of
>Celsius, the transform is easy enough but we need some way (in the config
>file) to tell the code the DesiredUnit is (in this example) Fahrenheit not
>Celsius

@jblance you responded with a higher level comment but I'll respond here. We can do that at the display level, the DesiredUnit could be Temperature that outputs C, F or K depending on config.